definition_validator/validator.cc

```diff
@@ -901,6 +907,27 @@ class ValidateWalk {
         }
     }
 
+    core::AutocorrectSuggestion suggestMethodDefinition(const core::GlobalState &gs,
+                                                        const core::Loc classOrModuleDefLoc,
+                                                        const core::MethodRef abstractMethodRef) {
+        auto [_, indentLength] = classOrModuleDefLoc.findStartOfLine(gs);
+        string indent(indentLength, ' ');
+
+        auto methodDefinition = core::lsp::prettyTypeForMethod(gs, abstractMethodRef, nullptr,
+                                                               abstractMethodRef.data(gs)->resultType, nullptr, true);
+
+        vector<string> reindentedMethodDefinitionLines;
+        absl::c_transform(absl::StrSplit(methodDefinition, "\n"), std::back_inserter(reindentedMethodDefinitionLines),
+                          [indent](auto &line) -> string { return absl::StrCat(fmt::format("{}  ", indent), line); });
+
+        auto reindentedMethodDefinition = absl::StrJoin(reindentedMethodDefinitionLines, "\n");
+
+        return core::AutocorrectSuggestion{
+            fmt::format("Define `{}`", abstractMethodRef.show(gs)),
+            {core::AutocorrectSuggestion::Edit(
+                {classOrModuleDefLoc.copyEndWithZeroLength(), fmt::format("\n{}", reindentedMethodDefinition)})}};
```

Comment:
This causes the fix to be inserted as the first line of the implementing class/module. That is a bit unfortunate, because I suspect it will often cause code to be inserted above `include` and `extend` statements, as well as class methods that commonly appear as the first few lines of code in class. Would it be better to insert at the bottom the class instead? Is there an easy way to do that? Of course that will have its drawbacks too (e.g. appearing below a floating `private` modifier), but perhaps it covers the common case better.

Reply:
> Would it be better to insert at the bottom the class instead? Is there an easy way to do that?

Yes, I think we likely want to implement this.

Luckily, we do actually have enough information at this point in Sorbet, though it's not great.

At the call to `validateAbstract`, we actually have both the `ast::ClassDef` node and the `core::ClassOrModuleRef`. The `ClassOrModuleRef`'s loc only stores the declLoc, which is the truncated first line of the definition. (You can see how the declLoc of a ClassDef is computed [here](https://github.com/sorbet/sorbet/blob/23b9ea187ae26d67a13a48b738b24a9e3da14586/parser/Builder.cc#L841).).

But the ast::ClassDef also carries the full enclosing location of the class def, from the `class` keyword to the `end` keyword. To find the last line of the class def, we can:

- double check that the ClassDef starts and ends on separate lines
- double check that the ClassDef's loc actually ends with an `end` keyword. It does not necessarily have to, because various Sorbet constructs will translate into class definitions, for example `X = Struct.new(:foo)` translates to `class X < Struct ... end`, so there is no `end` keyword in the source text. To check whether there was an `end` keyword, the easiest way is to do something [like this code](https://github.com/sorbet/sorbet/blob/23b9ea187ae26d67a13a48b738b24a9e3da14586/cfg/builder/builder_walk.cc#L620-L627). (While it would be nice to store more location information on the ClassDef and not have to rely on textual source parsing like this, Sorbet is under intense memory pressure on large codebases.)
- find the line of the `end` keyword, using the methods available in Loc.h to split a loc into its file+line combinations
- do some similar manipulations as you are already doing to craft the new replacement loc

In my opinion, this is definitely worth doing, because I think that abstract modules are more common than abstract classes.

From this, we also find some more good test cases for you:

- [test 1](https://sorbet.run/#%23%20typed%3A%20true%0A%0Aclass%20MyAbstractClass%0A%20%20extend%20T%3A%3ASig%0A%20%20extend%20T%3A%3AHelpers%0A%0A%20%20abstract!%0A%0A%20%20sig%20%7Babstract.void%7D%0A%20%20def%20foo%3B%20end%0Aend%0A%0AX%20%3D%20Class.new%28MyAbstractClass%29)
- [test 2](https://sorbet.run/#%23%20typed%3A%20true%0A%0Aclass%20MyAbstractClass%0A%20%20extend%20T%3A%3ASig%0A%20%20extend%20T%3A%3AHelpers%0A%0A%20%20abstract!%0A%0A%20%20sig%20%7Babstract.void%7D%0A%20%20def%20foo%3B%20end%0Aend%0A%0AX%20%3D%20Class.new%28MyAbstractClass%29%20do%0A%20%20def%20bar%3B%20end%0Aend)
- [test 3](https://sorbet.run/#%23%20typed%3A%20true%0A%0Aclass%20MyAbstractClass%0A%20%20extend%20T%3A%3ASig%0A%20%20extend%20T%3A%3AHelpers%0A%0A%20%20abstract!%0A%0A%20%20sig%20%7Babstract.void%7D%0A%20%20def%20foo%3B%20end%0Aend%0A%0Aclass%20OneLine%20%3C%20MyAbstractClass%3B%20end%0Aclass%20EmptyBody%20%3C%20MyAbstractClass%0Aend)

Reply:
One more thing to this effect: I recently came to this same conclusion while working on an unrelated autocorrect: to [re-declare a type member](https://sorbet.run/#%23%20typed%3A%20true%0A%0Aclass%20Parent%0A%20%20extend%20T%3A%3AGeneric%0A%20%20X%20%3D%20type_template%0Aend%0A%0Aclass%20Child%20%3C%20Parent%0Aend) required by a parent class or module.

I had tried factoring out some code to share, so that it would be easier to build that autocorrect, but I mistakenly chose to insert the code at the top of the class body instead of at the bottom:

https://github.com/sorbet/sorbet/pull/6958

I had been meaning to revisit the changes I landed in that PR and make it put the autocorrect at the bottom of the class body. Looking at that PR might be helpful at least to spark some ideas, or maybe even as a point to build the "insert before end" changes discussed above.

Reply:
I implemented the "insert at bottom" behaviour in 954d200.

Additionally, in an effort to support the common single-line stub pattern from the third test case you provided (i.e. `class Foo < Bar; end`), I changed how this autocorrect works somewhat significantly in 7231ed1. Now we provide a single autocorrect fix that defines all missing methods at once rather than providing separate fixes for each method.

This approach was inspired by the TypeScript LSP, which does something similar:

<img width="1007" alt="CleanShot 2023-08-14 at 10 29 50@2x" src="https://github.com/sorbet/sorbet/assets/776168/6dfb57d0-3102-43cc-a8ee-e2ec29f1d953">

<img width="311" alt="CleanShot 2023-08-14 at 10 30 41@2x" src="https://github.com/sorbet/sorbet/assets/776168/c76f507b-a670-4346-87d9-729ceacff02c">
<br/>
<br/>

Ultimately, this helps with the stub pattern because it gives us a way to dodge a bug that the previous approach was subject to when there were multiple missing method definitions and the user attempted to fix them all at once using the "Apply all Sorbet fixes for file" global fix. In that scenario, because the fix for the stub requires opening the method body to add new methods (as opposed to just adding new lines within a method body), individual fixes did not compose properly. That aside, I think the new behaviour is also a more user-friendly solution to the problem.

To recap where support for each of the three test cases above now stands:

- [test 1](https://sorbet.run/#%23%20typed%3A%20true%0A%0Aclass%20MyAbstractClass%0A%20%20extend%20T%3A%3ASig%0A%20%20extend%20T%3A%3AHelpers%0A%0A%20%20abstract!%0A%0A%20%20sig%20%7Babstract.void%7D%0A%20%20def%20foo%3B%20end%0Aend%0A%0AX%20%3D%20Class.new%28MyAbstractClass%29) is not supported, because [I wasn't sure how to consistently recognize all the permutations of a single-line class/module definition](https://github.com/sorbet/sorbet/blob/c5e74cc94e589f8440514b5ff6cf13c958100007/definition_validator/validator.cc#L991-L995) (as opposed to the common permutation described above). Open to feedback as to how to improve this!
- [test 2](https://sorbet.run/#%23%20typed%3A%20true%0A%0Aclass%20MyAbstractClass%0A%20%20extend%20T%3A%3ASig%0A%20%20extend%20T%3A%3AHelpers%0A%0A%20%20abstract!%0A%0A%20%20sig%20%7Babstract.void%7D%0A%20%20def%20foo%3B%20end%0Aend%0A%0AX%20%3D%20Class.new%28MyAbstractClass%29%20do%0A%20%20def%20bar%3B%20end%0Aend) is fixed, with additional test coverage.
- [test 3](https://sorbet.run/#%23%20typed%3A%20true%0A%0Aclass%20MyAbstractClass%0A%20%20extend%20T%3A%3ASig%0A%20%20extend%20T%3A%3AHelpers%0A%0A%20%20abstract!%0A%0A%20%20sig%20%7Babstract.void%7D%0A%20%20def%20foo%3B%20end%0Aend%0A%0Aclass%20OneLine%20%3C%20MyAbstractClass%3B%20end%0Aclass%20EmptyBody%20%3C%20MyAbstractClass%0Aend) is fixed, with additional test coverage.

Reply:
I wanted to comment on two other things that I think could use improvement here:

1. When several methods are missing in a class, the methods appear in the error message, and are also inserted into the source text, in an arbitrary order (see [this example](https://sorbet.run/#%23%20typed%3A%20strict%0A%0Aclass%20Parent%0A%20%20extend%20T%3A%3ASig%0A%20%20extend%20T%3A%3AHelpers%0A%20%20abstract!%0A%0A%20%20sig%20%7B%20abstract.void%20%7D%0A%20%20def%20foo%3B%20end%0A%0A%20%20sig%20%7B%20abstract.void%20%7D%0A%20%20def%20bar%3B%20end%0A%0A%20%20sig%20%7B%20abstract.void%20%7D%0A%20%20def%20qux%3B%20end%0Aend%0A%0Aclass%20Child%20%3C%20Parent%3B%20end), where the methods are declared from top to bottom as `foo`, `bar`, `qux`, but are listed in the error message as `foo`, `qux`, `bar`). This appears to be due to the fact that the [`members` method](https://github.com/lerebear/sorbet/blob/c5e74cc94e589f8440514b5ff6cf13c958100007/core/Symbols.h#L603), which is used [here](https://github.com/lerebear/sorbet/blob/c5e74cc94e589f8440514b5ff6cf13c958100007/definition_validator/validator.cc#L870), returns an unordered collection.
2. The error message contains some lines that describe the autocorrect edits which I think are unnecessary (see [this image](https://github.com/sorbet/sorbet/assets/776168/1af1d373-ea4e-4e43-a818-2422c334fcec)). I'm not sure why [we add those lines](https://github.com/lerebear/sorbet/blob/c5e74cc94e589f8440514b5ff6cf13c958100007/core/Error.cc#L188-L193), but they don't seem well suited to this particular context. The TypeScript error pictured above, which omits those lines, is easier to digest in my opinion.

Open to feedback as to whether or not either of those should be addressed immediately.

Reply:
> Open to feedback as to how to improve this!

The fact that you have detected this case and stopped short of suggesting an autocorrect which would do something like generate a syntactically invalid file after applying the autocorrect is good enough for me. It will always be possible to revisit that if/when there's a better idea for how to structure the autocorrect.

> I'm not sure why we add those lines

That's partly there's one interface for printing errors to the console and reporting errors to LSP.

In the command line interface to Sorbet, we need some way to tell the user what autocorrect they can expect to have happen when they re-run with the `-a` flag.

In LSP, I agree that those errors can be seen as clutter and redundant with both the `(fix available)` in the header and the code action title ("Define inherited abstract methods").

I'm happy to workshop how we present errors, but for the sake of reducing this PR's scope, I think I'd like to keep what currently exists. Once this PR lands, feel free to @mention me in the #internals channel on the Sorbet slack (https://sorbet.org/slack) and we can hash out some ideas for changes to make to how errors like these are presented.

> in an arbitrary order

Can we sort by their string names?

In general, we want all the output from Sorbet to be in a stable order, so that our snapshot tests are not flaky.

Reply:
> Once this PR lands, feel free to @mention me in the #internals channel on the Sorbet slack (https://sorbet.org/slack) and we can hash out some ideas for changes to make to how errors like these are presented.

Thanks, and will do 👍🏿 

> Can we sort by their string names?

Certainly: will fix.

Reply:
> Can we sort by their string names?

Fixed in f0996c9.